addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java

```diff
@@ -0,0 +1,310 @@
+                    && !command.toString().equalsIgnoreCase(UNKNOWN_USER_NAME) && (accessDeniedFace != null)) {
+
+                MatVector images = new MatVector(1);
+                images.put(0, accessDeniedFace);
+
+                // Setup label
+                Mat labels = new Mat(1, 1, CV_32SC1);
+                IntBuffer labelsBuffer = labels.createBuffer();
+                int label = command.toString().toLowerCase().hashCode();
+                labelsBuffer.put(0, label);
+
+                // Train face
+                faceRecognizer.update(images, labels);
+                faceRecognizer.setLabelInfo(label, command.toString());
+                faceRecognizer.save(MODELS_FILE);
+                logger.debug("Trained face of user \"{}\"", command.toString());
+            }
+        }
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Initializing {}::{}", getThing().getLabel(), getThing().getUID());
+        updateStatus(ThingStatus.UNKNOWN);
+
+        try {
+            // Read configuration
+            URI camera = new URI((String) getConfig().get("camera"));
+            String username = (String) getConfig().get("username");
+            String password = (String) getConfig().get("password");
+            interval = (BigDecimal) getConfig().get("interval");
+            threshold = (BigDecimal) getConfig().get("threshold");
+
+            // Load models file
+            File file = new File(MODELS_FILE);
+
+            if (file.exists()) {
+                faceRecognizer.load(MODELS_FILE);
+            }
+
+            // Load face cascade
+            faceCascade = new CascadeClassifier(FACE_CLASSIFIER_FILE);
+
+            // Add credentials for HTTP Digest and Basic authentication
+            AuthenticationStore auth = httpClient.getAuthenticationStore();
+            auth.clearAuthentications();
+            auth.addAuthentication(new BasicAuthentication(camera, Authentication.ANY_REALM, username, password));
+            auth.addAuthentication(new DigestAuthentication(camera, Authentication.ANY_REALM, username, password));
+            httpClient.start();
+
+            // Start making periodically images
+            job = service.scheduleAtFixedRate(() -> snapshot(camera), 0, interval.longValue(), TimeUnit.MILLISECONDS);
+
+        } catch (Exception e) {
+            logger.error("{}: {}", e.getClass().getName(), e.getMessage());
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+        }
+    }
+
+    @Override
+    public void dispose() {
+        logger.debug("Disposing {}::{}", getThing().getLabel(), getThing().getUID());
+
+        if (!job.isDone()) {
+            logger.debug("Cancelling snapshot job");
+            job.cancel(true);
+        }
+
+        try {
+            httpClient.stop();
+
+        } catch (Exception e) {
+            logger.error("{}", e.getMessage());
```

Comment:
Don't use the error level that is reserved to the framework. The last argument has to be the exception itself to get the stacktrace. The line should read: `logger.warn(e);`